### ISSUE TITLE
Unexpected error code returned by Vulkan

### ISSUE THREAD

**Issue by Limeth** [tomaka/vulkano#553]
I am getting an `internal error: entered unreachable code: Unexpected error code returned by Vulkan: 3294955296` on [this line](https://github.com/Limeth/vulkanscratch/blob/ee9066115838cbe7f280acd0fbf77bc8627fd21d/src/main.rs#L64) of my project, when I try to compile it.

Stacktrace:
```rust
$ env RUST_BACKTRACE=1 cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/vulkanscratch`
thread 'main' panicked at 'internal error: entered unreachable code: Unexpected error code returned by Vulkan: 3294955296', /home/limeth/.cargo/git/checkouts/vulkano-fcdd9024751cb884/934169d/vulkano/src/lib.rs:229
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:371
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:511
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:495
   7: vulkano::check_errors
             at /home/limeth/.cargo/git/checkouts/vulkano-fcdd9024751cb884/934169d/vulkano/src/lib.rs:229
   8: <vulkano::pipeline::compute_pipeline::ComputePipeline<Pl>>::with_unchecked_pipeline_layout
             at /home/limeth/.cargo/git/checkouts/vulkano-fcdd9024751cb884/934169d/vulkano/src/pipeline/compute_pipeline.rs:141
   9: <vulkano::pipeline::compute_pipeline::ComputePipeline<()>>::new
             at /home/limeth/.cargo/git/checkouts/vulkano-fcdd9024751cb884/934169d/vulkano/src/pipeline/compute_pipeline.rs:67
  10: vulkanscratch::WorkerDevice::new
             at src/main.rs:64
  11: vulkanscratch::main
             at src/main.rs:97
  12: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
  13: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:433
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  14: main
  15: __libc_start_main
  16: _start
```

**Comment by Limeth** [comment]
Also happens after updating to [the latest commit](https://github.com/tomaka/vulkano/commit/6c4aef673932c9c6dfa01a9edc1856efcbcbc94b): [updated repo](https://github.com/Limeth/vulkanscratch/tree/c15c825af07e5a2645b5608e199c26cb88dd5bcd).

**Comment by tomaka** [comment]
That code is `VK_ERROR_INVALID_SHADER_NV`, which isn't supported on purpose by vulkano.

**Comment by tomaka** [comment]
What this means is that you are probably trying to do something weird in your shader that glslang doesn't like.

**Comment by Limeth** [comment]
I think it might be caused by me trying to use `uint64_t` types in the shader using
```comp
#extension GL_ARB_gpu_shader_int64 : require
```

**Comment by Limeth** [comment]
No, that's not the cause. Ugh...

**Comment by Limeth** [comment]
Debug output to share with those who might come across the same issue.
```
    Finished dev [unoptimized + debuginfo] target(s) in 28.62 secs
     Running `target/debug/vulkanscratch`
DebugReport debug: Added callback
DebugReport debug: Added callback
DebugReport debug: Added callback
DebugReport debug: Added callback
ObjectTracker information: OBJ[0x2] : CREATE DebugReportCallbackEXT object 0x7fa098c21008
DebugReport debug: Added callback
DebugReport debug: Added callback
loader debug: Device Extension: VK_KHR_swapchain (libGLX_nvidia.so.0) version 0.0.68
loader debug: Device Extension: VK_NV_glsl_shader (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_KHR_sampler_mirror_clamp_to_edge (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_NV_dedicated_allocation (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_NV_external_memory (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_NVX_device_generated_commands (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_KHR_swapchain (Unknown) version 0.0.68
loader debug: Device Extension: VK_NV_glsl_shader (Unknown) version 0.0.1
loader debug: Device Extension: VK_KHR_sampler_mirror_clamp_to_edge (Unknown) version 0.0.1
loader debug: Device Extension: VK_NV_dedicated_allocation (Unknown) version 0.0.1
loader debug: Device Extension: VK_NV_external_memory (Unknown) version 0.0.1
loader debug: Device Extension: VK_NVX_device_generated_commands (Unknown) version 0.0.1
loader debug: Loading layer library libVkLayer_unique_objects.so
loader information: Insert device layer VK_LAYER_GOOGLE_unique_objects (libVkLayer_unique_objects.so)
loader debug: Loading layer library libVkLayer_swapchain.so
loader information: Insert device layer VK_LAYER_LUNARG_swapchain (libVkLayer_swapchain.so)
loader debug: Loading layer library libVkLayer_core_validation.so
loader information: Insert device layer VK_LAYER_LUNARG_core_validation (libVkLayer_core_validation.so)
loader debug: Loading layer library libVkLayer_object_tracker.so
loader information: Insert device layer VK_LAYER_LUNARG_object_tracker (libVkLayer_object_tracker.so)
loader debug: Loading layer library libVkLayer_parameter_validation.so
loader information: Insert device layer VK_LAYER_LUNARG_parameter_validation (libVkLayer_parameter_validation.so)
loader debug: Loading layer library libVkLayer_threading.so
loader information: Insert device layer VK_LAYER_GOOGLE_threading (libVkLayer_threading.so)
loader debug: Device Extension: VK_KHR_swapchain (libGLX_nvidia.so.0) version 0.0.68
loader debug: Device Extension: VK_NV_glsl_shader (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_KHR_sampler_mirror_clamp_to_edge (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_NV_dedicated_allocation (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_NV_external_memory (libGLX_nvidia.so.0) version 0.0.1
loader debug: Device Extension: VK_NVX_device_generated_commands (libGLX_nvidia.so.0) version 0.0.1
ObjectTracker information: OBJ[0x3] : CREATE Device object 0x7fa098d8d800
ObjectTracker information: OBJ[0x4] : CREATE VK_DEBUG_REPORT_OBJECT_TYPE_QUEUE_EXT object 0x7fa098c15a00
ObjectTracker information: OBJ[0x5] : CREATE Buffer object 0x1
ObjectTracker information: OBJ[0x6] : CREATE DeviceMemory object 0x2
ObjectTracker information: OBJ[0x7] : CREATE Buffer object 0x3
ObjectTracker information: OBJ[0x8] : CREATE DeviceMemory object 0x4
ObjectTracker information: OBJ[0x9] : CREATE CommandPool object 0x5
ObjectTracker information: OBJ[0xa] : CREATE VK_DEBUG_REPORT_OBJECT_TYPE_COMMAND_BUFFER_EXT object 0x7fa093341000
ObjectTracker information: OBJ[0xb] : CREATE Buffer object 0x6
ObjectTracker information: OBJ[0xc] : CREATE Buffer object 0x7
ObjectTracker information: OBJ[0xd] : CREATE ShaderModule object 0x8
ObjectTracker information: OBJ[0xe] : CREATE DescriptorSetLayout object 0x9
ObjectTracker information: OBJ[0xf] : CREATE PipelineLayout object 0xa
Driver error: Compute info
------------
Internal error: assembly compile error for compute shader at offset 0:
-- error message --
line 1, column 1:  error: invalid compute program header
-- internal assembly text --

ParameterValidation warning: vkCreateComputePipelines: returned VK_ERROR_INVALID_SHADER_NV, indicating that one or more shaders failed to compile or link
thread 'main' panicked at 'internal error: entered unreachable code: Unexpected error code returned by Vulkan: 3294955296', /home/limeth/.cargo/git/checkouts/vulkano-fcdd9024751cb884/6c4aef6/vulkano/src/lib.rs:229
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:371
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:511
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:495
   7: vulkano::check_errors
             at /home/limeth/.cargo/git/checkouts/vulkano-fcdd9024751cb884/6c4aef6/vulkano/src/lib.rs:229
   8: <vulkano::pipeline::compute_pipeline::ComputePipeline<Pl>>::with_unchecked_pipeline_layout
             at /home/limeth/.cargo/git/checkouts/vulkano-fcdd9024751cb884/6c4aef6/vulkano/src/pipeline/compute_pipeline.rs:141
   9: <vulkano::pipeline::compute_pipeline::ComputePipeline<()>>::new
             at /home/limeth/.cargo/git/checkouts/vulkano-fcdd9024751cb884/6c4aef6/vulkano/src/pipeline/compute_pipeline.rs:67
  10: vulkanscratch::WorkerDevice::new
             at src/main.rs:65
  11: vulkanscratch::main
             at src/main.rs:128
  12: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
  13: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:433
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  14: main
  15: __libc_start_main
  16: _start
ObjectTracker information: OBJ_STAT Destroy PipelineLayout obj 0xa (12 total objs remain & 0 PipelineLayout objs).
ObjectTracker information: OBJ_STAT Destroy DescriptorSetLayout obj 0x9 (11 total objs remain & 0 DescriptorSetLayout objs).
ObjectTracker information: OBJ_STAT Destroy ShaderModule obj 0x8 (10 total objs remain & 0 ShaderModule objs).
ObjectTracker information: OBJ_STAT Destroy Buffer obj 0x7 (9 total objs remain & 3 Buffer objs).
ObjectTracker information: OBJ_STAT Destroy Buffer obj 0x6 (8 total objs remain & 2 Buffer objs).
ObjectTracker information: OBJ_STAT Destroy CommandBuffer obj 0x7fa093341000 (7 total objs remain & 0 CommandBuffer objs).
ObjectTracker information: OBJ_STAT Destroy CommandPool obj 0x5 (6 total objs remain & 0 CommandPool objs).
ObjectTracker information: OBJ_STAT Destroy Buffer obj 0x1 (5 total objs remain & 1 Buffer objs).
ObjectTracker information: OBJ_STAT Destroy Buffer obj 0x3 (4 total objs remain & 0 Buffer objs).
ObjectTracker information: OBJ_STAT Destroy DeviceMemory obj 0x2 (3 total objs remain & 1 DeviceMemory objs).
ObjectTracker information: OBJ_STAT Destroy DeviceMemory obj 0x4 (2 total objs remain & 0 DeviceMemory objs).
ObjectTracker information: OBJ_STAT Destroy Device obj 0x7fa098d8d800 (1 total objs remain & 0 Device objs).
ObjectTracker information: OBJ_STAT Destroy Queue obj 0x7fa098c15a00 (0 total objs remain & 0 Queue objs).
loader debug: Unloading layer library libVkLayer_threading.so
loader debug: Unloading layer library libVkLayer_parameter_validation.so
loader debug: Unloading layer library libVkLayer_object_tracker.so
loader debug: Unloading layer library libVkLayer_core_validation.so
loader debug: Unloading layer library libVkLayer_swapchain.so
loader debug: Unloading layer library libVkLayer_unique_objects.so
```

**Comment by Limeth** [comment]
Updating to the latest nvidia driver (381) reveals more information about the error:
What follows
```
Driver error: Compute info
------------
(0) : error C6007: Constant register limit exceeded; more than 1024 constant registers needed to compiled program
```
are 12 000 repetitions of this line:
```
(0) : error C6007: Constant register limit exceeded; more than 1024 constant registers needed to compiled program
```
followed by
```
(0) : error C6020: Constant register limit exceeded at infinity; more than 1024 registers needed to compiled program
(0) : error C6020: Constant register limit exceeded at prec; more than 1024 registers needed to compiled program
```

It looks to me like an infinite loop of sorts is generating these errors. Odd.

**Comment by tomaka** [comment]
You should try again after #556

**Comment by Limeth** [comment]
Thanks, but unfortunately, no luck.

**Comment by Limeth** [comment]
Just an update; I received a response from the NVIDIA Developer Support:
> Our engineers from Vulkan team have seen something like this when the shader uses too large of a UBO. Our limit is 64KB. It looks like you have a single UBO with a huge array of structure in it.

**Comment by tomaka** [comment]
Ah! Vulkano doesn't check the limits of shaders yet, so it could be the problem.

**Comment by tomaka** [comment]
Closing in favour of #585
